### ISSUE TITLE
localStorage is spammed with state/nonce combos

### ISSUE THREAD

**Issue by csvn** [auth0/auth0.js#402]
Every time before `auth0.js` performs a redirect to the `/authorize` endpoint, `state` and `nonce` are generated and stored in `localStorage` for verification upon return. If the user lands on the Auth0 hosted login page but does not login, the stored `state/nonce` combo won't be removed.

This means that `localStorage` can fill up with lots of unused values, especially during development.
```
com.auth0.auth.3z-j16G6OFMF_v-nt_QUZS5VV~vf.fGD:	{"nonce":"55yFu~kZUv0KT6azRtCIyf_c3dx-GAmJ"}
```
![image](https://cloud.githubusercontent.com/assets/8770194/24415185/ec2b9ea8-13e0-11e7-94d3-8b26c2d499c0.png)

Is there a reason why this could not be done via a single `localStorage` key, with JSON data?
Or possibly that in addition to `nonce`, a `issuedAt` field could be used for cleaning up old `state/nonce` combos?

**Comment by luisrudge** [comment]
This is indeed inconvenient. I added this to our backlog and will tackle in the future.

**Comment by csvn** [comment]
If it's something that you would appreciate a pull request on, you can mention it here. Not sure if there should be requirements about how long to store `state/nonce` items, if it should be configurable, etc.

**Comment by luisrudge** [comment]
I'd love a PR. The thing is that I don't know what we should do in this case. What do you propose? add a `issuedAt` and clean the expired ones on start?

**Comment by csvn** [comment]
Yeah, I think something like that might be fine for now.

1. Add a `issuedAt` field for the JSON object where `nonce` is stored
  ```
  "com.auth0.auth.<state>": "{ nonce: <nonce>, issuedAt: 1490783455 }"
  ```
2. (Possibly) Added a configurable field or default value to `auth0.js` options, for the number of hours/days before `state/nonce` pairs are removed
3. During `auth0.js` initialization, loop through localStorage keys for those prefixed with the namespace (`com.auth0.auth`), and remove outdated `state/nonce` pairs

**Comment by csvn** [comment]
It would be possible to store all `state/nonce` pairs in a single namespace too, but it would require `JSON.parse` and `stringify` for each read/write. This due to the fact that `auth0.js` might be used from multiple browser tabs at the same time.

**Comment by luisrudge** [comment]
Yeah. I like that. Let me talk with some people internally and I'll get back to you.

**Comment by bragma** [comment]
Any news on this? Thanks!

**Comment by hzalaz** [comment]
Sadly no news, we are busy working on other bugs/features in our SDKs.

**Comment by bragma** [comment]
@hzalaz ok thanks, is there any workaround I can used in my own code to cleanup?

**Comment by trickkiste** [comment]
I am seeing this too!

Quite annoying. I am using webAuth.renewAuth() to silently reauthenticate when I return to the page and still have a valid sso session. In my dev env, I am doing that every 30 seconds.

I got 12000 nonce pairs in storage right now.

Also webAuth.renewAuth() in combination with React does only work after `document.readyState === "complete"`
```
      document.onreadystatechange = () => {
        if (document.readyState === "complete") {
          console.log("document.readyState === complete");
          this.renewUserAuth();
        }
      }
```

**Comment by oliverlloyd** [comment]
Predictably, this problem is causing LS to fill up for us and we start to get QuotaExceeded errors:

`DOMException: Failed to execute 'setItem' on 'Storage': Setting the value of 'com.auth0.auth.A-XXXX.~sh9JOAJ~XXXvljHCX0s1xXXX' exceeded the quota.`

At this point things start to go badly wrong... All setItem calls, not just auth0's, start to fail. And it only seems to take a couple of weeks of steady usage to reach this point.

This bug seems like a bit of a ticking time bomb rather than an inconvenience. I can't expect users to clear their cache every few weeks but unless they do this the app will break. Any chance this issue can be reviewed?

My preference would be to use a single key. Mainly because code like this:
```
for (var key in localStorage){
    // Check prefix of key
}
```
can be very slow when you have lots of keys...

To handle multiple tabs or asynchronous calls, why not stringify an array of objects?

```
"com.auth0.auth": "[
  { state: <state>, nonce: <nonce> },
  { state: <state>, nonce: <nonce> },
  { state: <state>, nonce: <nonce> },
  ...
]"
```
Then, to prevent overflow, each time you push() to it you can check this array's length and shift() as required. A max length of 50 should do it.

Sure, you'd need to parse/stringify each time and any nonce checks would need to loop this array checking for a state match but this seems like the correct way to implement this solution.

**Comment by bentedder** [comment]
Any news on this item? I'd be happy to take a look at providing a solution if somebody can point me in the right direction.

**Comment by luisrudge** [comment]
This only happens when an auth request fails, right? after you successfully log in, the key should be removed from localStorage. Is that the case for you?

**Comment by aguerere** [comment]
@luisrudge I am still having this issue when using the "silent" renewal for the implicit flow. Does this depend on the auth0js version? I am currently using v8.8 .

**Comment by luisrudge** [comment]
Do you get this only when an auth request fails? or even for successful auth requests?

**Comment by luisrudge** [comment]
We remove the item from localStorage here: https://github.com/auth0/auth0.js/blob/7f7c586b664f67834fdebd810acc740002f80101/src/web-auth/index.js#L168

So, if you're calling parseHash, it should be removing the item. Are you calling parseHash?

**Comment by csvn** [comment]
@luisrudge This may also occur if the login attempt is canceled after redirect. Since they are not removed then they may fill up localstorage eventually.

**Comment by luisrudge** [comment]
yes, that's what I said [here](https://github.com/auth0/auth0.js/issues/402#issuecomment-330726482). Basically, it piles up when the auth request fails. We'll get to this issue eventually, but we are doing a big push on other areas right now. If someone wants to tackle this, I'll happily discuss and review a PR.

**Comment by aguerere** [comment]
We are calling parseHash and the entries in localStorage pile up even for successful logins.

**Comment by luisrudge** [comment]
@aguerere that's very strange. can you setup a repro project so I can debug it?

**Comment by seiyria** [comment]
While it's not a minimum repro project, [my project does this as well](https://github.com/LandOfTheRair/landoftherair).

**Comment by seiyria** [comment]
Left my project running overnight and woke up to 72k errors in my console about renewAuth failing to remove a DOM node (I had to add a function that, after auth, I clear all iframes on the page myself), as well as a warning that my localstorage was full. This is a bit more than inconvenient if it never cleans up after itself, that just means it's a ticking time bomb.

**Comment by seiyria** [comment]
In the meantime, my solution is this:

```
  private clearOldNonces() {
    Object.keys(localStorage).forEach(key => {
      if(!startsWith(key, 'com.auth0.auth')) return;
      localStorage.removeItem(key);
    });
  }
```

Having localStorage get full of nonces caused several problems:

- the aforementioned storage issues
- cant find anything manually in localstorage
- cant scroll through localStorage in the dev tools - too many items makes it slow and unresponsive

a function like this called on login and token renewal should suffice for keeping localStorage clean.

**Comment by nirmaldeol** [comment]
Hi i think in logout method you can do something like this.
This method is nearly same as above but that seems like a wrong implementation:
`private clearOldNonces() {
    Object.keys(localStorage).forEach(key => {
      if(!key.startsWith('com.auth0.auth')) return;
      localStorage.removeItem(key);
    });
  }
`

then call this from logout method

**Comment by chris31389** [comment]
@nirmaldeol if we can call that when it finds an expired token and no refresh token / expired refresh token

**Comment by nirmaldeol** [comment]
@chris31389 I am not sure if understood your question but yes we can do that under 
' isAuthenticated() ' method after checking token expiry.  And then calling above method with if() Logic

**Comment by ericmwalsh** [comment]
Bump... this is still an issue.

**Comment by daggett206** [comment]
+1, I am using auth0-lock and on 'authenticated' event

**Comment by rmariuzzo** [comment]
Just found that issue... would be happy to do a PR, can someone point me to the right direction?

**Comment by luisrudge** [comment]
@rmariuzzo Hi. What do you propose to fix this? We had a few suggestions in the thread, but I'm still not convinced they're the best ones.

**Comment by rmariuzzo** [comment]
@luisrudge I will have to dive into the code and see how it is used. If you can explain me a little the usage of the nonce I could come with a solution. 

---

I chimed in this thread because I was surprised on the amount of entries in my localStorage.

**Comment by luisrudge** [comment]
nonce and state are just arbitrary strings that you can provide (or we generate a random one). We send those strings to the server and check them when the server sends the response back to make both req/res match. state is also used to build the localStorage key:

`com.auth0.auth.state:	{"nonce": nonce}`

**Comment by ericmwalsh** [comment]
Hey guys just wanted to chime in with some feedback.  I've been experimenting with solutions to get rid of the extra nonces (I noticed a ton of them in my localStorage, probably similar to @rmariuzzo).  I had not yet run into issues w/ localStorage being filled up by them but seeing a stack of 20 was worrisome enough so I tried some of the solutions listed above...

```
  private clearOldNonces() {
    Object.keys(localStorage).forEach(key => {
      if(!startsWith(key, 'com.auth0.auth')) return;
      localStorage.removeItem(key);
    });
  }
```

While this can work you have to make sure you do not clearOldNonces at the wrong time... As @luisrudge mentioned the nonce is used as part of the server request.

Is it possible to add a unix timestamp to the nonces?  An extraneous field that will be used for nonce cleaning and is not part of the request.

(Just an idea/suggestion!)

**Comment by rmariuzzo** [comment]
@luisrudge sounds like we should clear the nonce after the server responds back. Isn't it?

**Comment by luisrudge** [comment]
@rmariuzzo indeed it should, and we are: https://github.com/auth0/auth0.js/issues/402#issuecomment-331653733 I'm not sure in what kind of scenarios we are not cleaning up them.

**Comment by rmariuzzo** [comment]
🤔 💭  then I will check later, when I get some time.

**Comment by mattiafioraso** [comment]
@luisrudge I noticed that this inconvenience shows up while performing a `checkSession()` when no session is actually active (so that the API call returns "unauthorized")

**Comment by bragma** [comment]
Same problem here. I am using the hosted login page. In my SPA:
1. Check for any valid token
2. Fail -> Call SDK authorize
3. Redirected to the hosted login page, enter valid credentials
4. Redirected to my SPA, I call parseHash
5. I get the tokens, but the state/nonce is still there

Is there anything I can do to help debugging?

**Comment by bragma** [comment]
After a quick investigation, I found that the entire cleanup process is very brittle.
Methods creating transactions (i.e. nonces) are:

renewAuth
checkSession
passwordlessStart
authorize
passwordlessVerify

The problems is that session are ONLY removed on a successful parseHash AND if the state has not been tampered with. A lot of things can go wrong. Few examples follow:

1. User reloads, interrupts, somehow breaks an authorization flow requiring a redirect: parseHash will not find out required data and the session will not be removed. Example:

From your app, call authorize -> redirect to hosted login page -> do not login, reload the app -> state is lost, nonce will not be removed

2. App calls checkSession, something fails in between. checkSession has this flow:

checkSession()
...
create a transaction
...
call webMessageHandler.run -> fail (ex: consent_required)  -> cb(err) -> stale transaction

I suspect there are a lot of other flows that can fail. just checked the one I use most.

**Comment by luisrudge** [comment]
We can only clean the storage keys when we have a valid state, since we use the state as part of the key. If the state has been tampered or something else happens, we can never match the keys.

**Comment by bragma** [comment]
@luisrudge I understand, still this causes entries to pile up. What do you think of something like this?

	function _clearOldNonces(expirationSeconds) {
		var now = _currentTime();
		var expirationTime = now - expirationSeconds * 1000;
		Object.keys(localStorage).forEach(function(key) {
			if(!key.startsWith('com.auth0.auth')) return;
			var el = JSON.parse(localStorage.getItem(key));
			if (!el.sweepTime) {
				el.sweepTime = now;
				localStorage.setItem(key, JSON.stringify(el));
			} else if (el.sweepTime < expirationTime) {
				localStorage.removeItem(key);
			}
		});
	}

And leaving a 1 hour window?

**Comment by luisrudge** [comment]
We're looking into alternatives here. The idea was to use store-js with the expire plugin (https://github.com/marcuswestin/store.js/#list-of-all-plugins), but turns out it has poor support for safari in private mode (something we also experience in https://github.com/auth0/auth0.js/issues/623) and the library doesn't look like it is being maintained anymore, so we're back to the drawing board. I like your idea, though. We'd just need to set a `sweepTime` when we create a transaction and clean all transactions when `WebAuth` is instantiated.

**Comment by bragma** [comment]
@luisrudge please consider if cleanup on instantiation is enough. In some case (thinking at SPA), WebAuth may be instantiated only once at application start, but several calls might be done without reloading and causing entries to pile up. It may not be enough, at the moment I am using that code before any call to WebAuth methods, it may be too much but I am trying to play it safe.
I am also using a 1 hour lifetime, which may not be a sensible value.

**Comment by luisrudge** [comment]
I think 1h is more than enough. We can probably reduce this amount as well, considering this state calls are meant to be used right away. I'll think about the instantiation thing. Maybe before each method call is too much as well so.. we'll see 🎉

**Comment by DaveTaddei-CR** [comment]
Also experiencing this since upgrading to auth0-lock v11 from v9 due to the upcoming deprecations

**Comment by danielmcardoso** [comment]
Same problem here. I'm new to Auth0 and in the trial to decide if our company adopts it, but during my prototype development found this problem. Trying auth0-lock and angular-jwt, latest stable versions. When I log in, localstorage entries are created and not removed. Worried about the impacts of this in the future and not sure on how to workaround this...

**Comment by luisrudge** [comment]
@danielmcardoso how are you using auth0? what methods etc?

**Comment by danielmcardoso** [comment]
@luisrudge, as I am still learning / testing auth0, what I am doing is quite simple, based on a few quick starts and how-tos. I have an AuthService class that has an instance of Auth0Lock, DI for JwtHelperService and methods for login(), logout() and check if currently isAuthenticated(). I inject this AuthService in AppComponent, where I have calls to login(), logout() and isAuthenticated(). That's all. Each time I call login() and then logout() later, new entries are added by auth0 to the localstorage and the old entries are kept, in a similar behavior to what others have described in this thread.

Here is the AuthService class:

```typescript

import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { environment } from 'environments/environment';

import { JwtHelperService } from '@auth0/angular-jwt';
import Auth0Lock from 'auth0-lock';

@Injectable()
export class AuthService {
    auth0Options = {
        auth: {
          redirectUrl: environment.auth0.callbackUrl,
          responseType: 'token id_token',
          audience: `https://${environment.auth0.domain}/userinfo`,
          params: {
            scope: 'openid profile'
          }
        },
        autoclose: true,
        oidcConformant: true,
      };

      lock: Auth0Lock = new Auth0Lock(
        environment.auth0.clientId,
        environment.auth0.domain,
        this.auth0Options
      );

    constructor(private jwtHelper: JwtHelperService) {
        this.lock.on('authenticated', (authResult: any) => {
            console.log('Nice, it worked!');
            this.lock.getUserInfo(authResult.accessToken, (error, profile) => {
                if (error) {
                  throw new Error(error);
                }
                localStorage.setItem(environment.auth0.tokenStorageName, authResult.idToken);
                localStorage.setItem('profile', JSON.stringify(profile));
            // this.router.navigate(['/']); // go to the home route
            });
        });

        this.lock.on('authorization_error', error => {
            console.log('something went wrong', error);
        });
    }

    login() {
        this.lock.show();
    }

    logout() {
        localStorage.removeItem('profile');
        localStorage.removeItem(environment.auth0.tokenStorageName);
    }

    isAuthenticated(): boolean {
        return !this.jwtHelper.isTokenExpired();
    }

}
```

**Comment by danielmcardoso** [comment]
@luisrudge, I tried with your page and it did work as expected (nonce items in localstorage were not kept). Testing a little bit more, I could stop the nonce items from being kept in localstorage in my app by simply adding the "email" scope param. I don't understand why this is necessary, but this seems to be the right direction in finding out what's happening.

**Comment by danielmcardoso** [comment]
I tested it. And it indeed worked (keys were cleared). I then made the same change to my app, removing the `email` scope and keeping only `openid profile` and... it worked as well! I managed to reproduce the original behavior (keys not cleared) by revoking the authorized client using the auth0 management > users feature. So maybe it has something to do not with having the `email` scope param, but with having it granted to the user?

**Comment by luisrudge** [comment]
what do you mean by revoking the authorized client?

**Comment by danielmcardoso** [comment]
Go to auth0 management UI, click Users, click on johnfoo in the list to open User Details, click on the Authorized Clients tab. You will see your client listed. If you click on the arrow near the Revoke button, it will show the granted scopes. Click the Revoke Button to revoke the client authorization for this user, including the granted scopes. When you try to login again using your app, it will ask for authorization. Make sure the email scope is not requested nor granted.

**Comment by luisrudge** [comment]
@danielmcardoso thank you very very much for your patience. I could reproduce this issue by using a custom audience. Using `email` or not in the scope didn't make a difference for me.

@danielmcardoso @DaveTaddei-CR  Are you both using a custom audience?

**Comment by danielmcardoso** [comment]
@luisrudge, what do you mean by "custom audience"? Could you please elaborate a bit on this concept? In the API settings (in auth0 management UI), I have filled the Identifier field with some text (actually a URL that identifies my API) and I use that text to fill the audience property of the options parameter of auth0lock constructor.

**Comment by luisrudge** [comment]
that's it :) using a custom `audience` parameter. If I add the `audience` option, it doesn't clear localstorage. If I remove it, it clears accordingly. Can you try that in your project as well (just for testing)?

**Comment by danielmcardoso** [comment]
I tried it without the custom audience and got the same results as you did (add `audience` option, doesn't clear localstorage. Remove `audience` option, keys are cleared as expected). I also confirmed again what I found out earlier about the `email` scope. If the user has this scope previously set in the Authorized Client Scopes, keys are cleared even if I add a custom `audience`. Could you re-check this, please?

**Comment by luisrudge** [comment]
Yup, that's it. I'll start investigating this in the next couple of days. I have some IE issues that have a bit more priority than this 😭

**Comment by luisrudge** [comment]
@danielmcardoso I can't repro this anymore (with or without custom audience). Can you write a step-by-step on how you can reproduce this so I can investigate?

**Comment by luisrudge** [comment]
@danielmcardoso I had some time to go back to this and I think I found the issue. Just so we can clarify: You're only seeing this with the google connection, right? Also, do you mind adding `rememberLastLogin: false` to your lock options and see if this still happens?

**Comment by DaveTaddei-CR** [comment]
@luisrudge I am using Google Idp, I set the flag as suggested but it still accumulates nonce values apparently without end.

**Comment by luisrudge** [comment]
@DaveTaddei-CR where did you set the flag?

**Comment by luisrudge** [comment]
@DaveTaddei-CR what are your lock options?

**Comment by DaveTaddei-CR** [comment]
autoclose: true,
    rememberLastLogin: false,
    auth: {
      redirectUrl: AUTH_CONFIG.callbackUrl,
      responseType: 'token id_token',
      audience: `https://${AUTH_CONFIG.domain}/userinfo`,
      params: {
        scope: 'openid app_metadata'
      }
    }

This was applied to the auth0-react-sample/01-Embedded-Login example.  It seems to be creating a nonce value and then removing again, all nonce values stored from before the change remain.  If i try different logins it records a nonce per login email which never goes away.

**Comment by luisrudge** [comment]
Previous nonces won't be removed. Are you saying that every new login is adding a new localStorage entry and not cleaning up?

**Comment by leoyli** [comment]
I like the idea of using short-living cookie, which seems to much efficient.  Right now I just clean up manually by a key loop after the authentication succeed, which indeed, inconvenient...

**Comment by lwensveen** [comment]
For me this is also an issue, I'm building a solution for the nonce/state mismatch when someone opens a magic link in outlook or some other mail browser.

The nonce/state is missing then since it's a different browser. My solution was to pull an email address from my token and immediately send a new magic link email with an instruction to use the same browser.

This works fine for once, but if someone logs in a month later they will have the nonce/state mismatch again because the nonce/state isn't being cleared properly.

**Comment by BrianMikinski** [comment]
@hzalaz @luisrudge is there any danger in deleting/clearing out the nonces after a user logs in?

**Comment by luisrudge** [comment]
@BrianMikinski no. they're used only to keep the state between authentication requests (sign in or token renewal)

**Comment by bill-vatomic** [comment]
I am seeing this issue as well. My localStorage is filled with multiple entries for "com.auth0.auth.<value>" and "co/verifier/<my custom domain>/<value>" each time a user logins in to my app. My application uses its own login page that calls the login() method on the Auth0 Web SDK (v9). Is this not the recommended usage for the SDK? I would rather not have to manually cleanup the localStorage.

**Comment by luisrudge** [comment]
@bill-vatomic can you confirm that after a successful login, those keys are *not* being removed?

**Comment by bill-vatomic** [comment]
@luisrudge yes, after a successful login (I get back access_token and id_token), there is still an entry in the localStorage called "co/verifier/<custom domain>/33f3-akF4L88" that is not deleted. Each subsequent login adds a new entry with a different value at the end.

**Comment by luisrudge** [comment]
all browsers? This should be removed on successful logins, so maybe there's an issue there.

**Comment by bill-vatomic** [comment]
Yes, in Firefox, Chrome, and Safari, I am seeing the repeated entries for "co/verifier/".

**Comment by luisrudge** [comment]
Do you have all of them with 3rd party cookies disabled?

**Comment by bragma** [comment]
As a side note, I am using universal login page on the original Auth0 domain. I've noticed that there are stale nonces on that page too. So they pile up in xxxx.eu.auth0.com too, with name "com.auth0.auth.-SOMETHING"

**Comment by GrEg00z** [comment]
As @bragma said, I can confirm while using universal login, after each authentification on the **auth0 hosted page** (https://<my project>.eu.auth0.com/login...), this page will keep in localstorage all  com.auth0.auth.xxxxx  previous entries, without doing any cleaning.

**This point must be considered in priority, because we can't do anything from our side to fix this.**

**Comment by luisrudge** [comment]
We have a fix for this here: https://github.com/auth0/auth0.js/pull/817
We're just evaluating possible undesired outcomes with this. I'll probably release this under a beta version so you can test that there are no side effects. The idea here is to use a short-lived cookie instead of localStorage.

**Comment by GrEg00z** [comment]
Ok so let me know when the beta will be released, I will test with pleasure

**Comment by bill-vatomic** [comment]
Thank you Luis! Let me know if you need help testing.

**Comment by luisrudge** [comment]
Hi friends! 
I have a beta version available for testing: https://cdn.auth0.com/js/auth0/9.7.4-beta1/auth0.js

In this version, we changed the default storage layer to be cookies instead of localStorage. This allows us to set a short expiration time for cookies and never worry about this kind of thing anymore. Can you please test this and make sure everything is still working? I already tested in a few test projects and Auth0 Quickstarts, but I want to make sure that this gets the necessary exposure before releasing it. Thanks!

cc @GrEg00z @bill-vatomic

**Comment by bill-vatomic** [comment]
Hi Luis!

I just gave this a shot and it seems to be working perfectly! I'm no longer seeing the duplicated entries in localStorage, just the expected id, access_token, and expiration.

Thank you for fixing this! Do you have any idea when this will roll into the main release? We will continue using the beta in the meantime to ensure no other issues, but this is looking good.

**Comment by luisrudge** [comment]
@bill-vatomic just a check: nothing will go to localStorage anymore, because we're using cookies by default. Check your cookies, see if they have a short expiration date etc. 
I'm glad the normal authentication flow works though!

**Comment by bill-vatomic** [comment]
Yes, I am seeing everything in cookies, with short expiration times. What I meant to say was I'm only seeing entries in localStorage that I am placing there myself.

**Comment by luisrudge** [comment]
@bill-vatomic ahhhh, perfect! Thanks for clarifying that. I'll wait a couple of days before releasing this into the wild.

**Comment by FBloembergen** [comment]
Hi Luis,

Will this solution also honor the custom domains you can set in auth0? Else some users will still experience problems when blocking third party cookies in their browser.

**Comment by luisrudge** [comment]
@FBloembergen everything set in the same page by auth0.js is considered a 1st party cookie, so that's not an issue.

**Comment by bragma** [comment]
@luisrudge my Auth0 hosted login page directly loads lock 11.9.0 from your CDN. Which version of lock should I use to fix this? Has it already been updated?
Thanks!

**Comment by luisrudge** [comment]
@bragma final release will happen next week. We've been powering through some hard issues in the last couple of weeks, but next week I'll make sure we get a new release of auth0.js and lock with this update. Thanks for the patience!

**Comment by lwensveen** [comment]
This might be a bit offtopic, but another issue I've noticed when a state mismatch occurs is that Auth0 logs it as an successful login attempt.

**Comment by luisrudge** [comment]
@lwensveen that's correct. The state checking happens after the authentication. It's a client side verification. As far as the server is concerned, the authentication happened successfully.

**Comment by andreasvirkus** [comment]
We have a Chrome extension which injects our app in an iframe for our users - this could result in a lot of instances being spun up rapidly so the nonce cookies were still an issue for us as they weren't cleared up fast enough. If anyone runs into a similar edge case, this is what we used to clear them up:

```javascript
export const removeAuth0TemporaryCookies = () => {
  document.cookie.split(';').forEach((cookie) => {
    const name = cookie.trim().split('=')[0]
    
    if (name.startsWith('_com.auth0.auth.') || name.startsWith('com.auth0.auth.')) {
      document.cookie = `${name}=; Domain=${location.hostname}; Path=/; Secure; SameSite=None; Expires=${new Date(0).toUTCString()}`
    }
  })
}


export const getSession = () => {
  return new Promise<GetSessionResponse>((resolve, reject) =>
    auth0.checkSession({}, (err, result) => {
      if (err) return reject(err)
      removeAuth0TemporaryCookies()
      return resolve(result)
    }),
  )
}
```
